### PR TITLE
[SYCL] Fix handling of multiple usages of composite spec constants

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Fixed the issue that instead of re-using previously assigned IDs
for elements of a composite spec constant, all elements used the
same ID, which was taken from the last element of the composite.

### CODE REVIEW

File: sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      auto sycl_kernel = p.get_kernel<kernel_t<T>>();
+      cgh.parallel_for(sycl_kernel, cl::sycl::range<1>(1), func);
+    });
+    q.wait();
+  }
+};
+
+template class kernel_driver_t<float>;
+
+// Uncomment this to trigger the bug.
```
Comment (dm-vodopyanov) [inline sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp:65]: Should the line 66 be commented? Or should the line 65 be removed?

Comment (AlexeySachkov) [inline sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp:65]:
```suggestion
// The line below instantiates the second use of a spec constant
// named `sc_kernel_t`, which used to corrupt the spec constant content
```

Comment (AlexeySachkov) [reply sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp:65]: 66 shouldn't be commented or the bug won't be triggered. But I agree that the comment can be improved - I posted a suggestion, please take a look

Comment (dm-vodopyanov) [reply sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp:65]: Thanks, looks good to me.

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 2]
```diff
@@ -336,21 +334,20 @@ Instruction *emitSpecConstantComposite(Type *Ty,
 Instruction *emitSpecConstantRecursiveImpl(Type *Ty, Instruction *InsertBefore,
                                            SmallVectorImpl<unsigned> &IDs,
                                            bool IsNewSpecConstant,
```
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:336]: This is no longer used - please remove

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp]: Fixed in e09b1b859bd88f8dc46b0aaad8e4bc21a7993b27

Comment (AlexeySachkov) [reply sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp]: Applied in e09b1b859bd88f8dc46b0aaad8e4bc21a7993b27

File: sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// UNSUPPORTED: cuda
+//
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %RUN_ON_HOST %t.out | FileCheck %s
+// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
+// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
+//
+// The test checks that multiple usages of the same specialization constant
+// works correctly with: toolchain processes them correctly and runtime can
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply sycl/test/on-device/spec_const/multiple-usages-of-composite.cpp]: Fixed in e09b1b859bd88f8dc46b0aaad8e4bc21a7993b27

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 2 above)
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/SpecConstants.cpp]: Removed in e09b1b859bd88f8dc46b0aaad8e4bc21a7993b27

File: llvm/test/tools/sycl-post-link/multiple-composite-spec-const-usages-2.ll [hunk 4]
```diff
@@ -0,0 +1,1450 @@
+; RUN: sycl-post-link -spec-const=rt --ir-output-only %s -S -o - \
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/test/tools/sycl-post-link/multiple-composite-spec-const-usages-2.ll:1]: Simplified the test in 7de5a3a25fb1fc80d95eaab96e97db2df53e2177

**Review by dm-vodopyanov — APPROVED** [review]
LGTM

File: llvm/test/tools/sycl-post-link/multiple-composite-spec-const-usages-2.ll (hunk 4 above)
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/test/tools/sycl-post-link/multiple-composite-spec-const-usages-2.ll:1]: Thanks, looks much better now. Nit: `@llvm.lifetime.start.p0i8` could be removed as well

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**